Refactor Datadog::HTTPTransport

delner opened the pull request:
The `Datadog::HTTPTransport` class is quite large and contains a lot of responsibility, which makes it difficult to test, and to extend.

The goal for this pull request is to refactor the `HTTPTransport` class to accomplish a few things:

 - **Make the transport simpler**: Break it into smaller pieces so those pieces are easier to understand and use.
 - **Make it easier to test**: By virtue of having smaller pieces, it should be easier to test the behavior of discrete parts.
 - **Make it easier to use in tests**: Some tests want to stub requests to the agent completely, others want to mock and inspect the expected HTTP output. By substituting in a test component at the last step of the transport, we can accomplish both of these goals while still running a maximal amount of the typical code paths.
 - **Make it easier to extend**: As we develop improved means of sending data to the agent (e.g. Keep-Alives, connection pooling, alternative HTTP libraries, etc), having a transport layer made of simpler, interchangeable, extendable parts will make implementation and maintenance of such strategies easier.
 - **Keep it performant**: Maintain or improve performance vs the old transport.

What's left to do:

 - [x] Add new `Transport::HTTP` components
 - [x] Add a compatibility layer to new transport (so it can be interchanged with the old)
 - [x] Add complimentary test components for new transport
 - [x] Run performance comparison of old vs new transport
 - [x] Update Writer to use new transport
 - [x] Update old tests
 - [x] ~~Add metrics from #587~~ (Won't do; will revisit this later)
 - [x] Add unit tests
 - [x] Deprecate old transport

delner commented:
https://github.com/DataDog/dd-trace-rb/pull/738 removed services from the trace library. This PR should be rebased on those changes, and the corresponding service components removed.

delner commented:
Rebased this PR on `0.23-dev` and removed services from it. Seems to be working okay.

brettlangdon commented on lib/ddtrace/transport/http/client.rb:
```diff
@@ -0,0 +1,94 @@
+          @apis = options[:apis] || DEFAULT_APIS
+
+          # Select active API
+          @active_api = apis[options.fetch(:api_version, V3)]
```
we should default to v4

delner replied:
:+1:

brettlangdon commented on lib/ddtrace/transport/http/client.rb:
```diff
@@ -0,0 +1,94 @@
+          raise UnknownApiVersion if active_api.nil?
+
+          # Select encoder
+          @encoder = options[:encoder]
```
don't we want to grab this from the `@active_api`?

delner replied:
I'll have to look at this again.

delner replied:
From what I can tell, I think this suggestion would be fine. I don't remember why I exactly wrote it this way: maybe because I was reworking what was in the old transport, maybe I wanted some kind of override (much of the code seems to allow for encoder to be overridden), but I don't remember exactly. Either way, it should be safe to set it to whatever the `active_api` is.

delner replied:
Okay, actually can't do this, because API doesn't define encoding, the endpoint does (which makes sense, since different endpoints can have encodings.)

I think it'd be better to consider if we need to force encoders on endpoints from the client (which is essentially the purpose it serves.) If not, we can just eliminate this option entirely from `Client`, have it be handled only in `ParcelEndpoint`, which would probably simplify the code without sacrificing much.

delner replied:
This is obsolete. Encoding has been incorporated into the `Datadog::Transport::HTTP::Traces::API::Endpoint` which defines how requests are prepared.

brettlangdon commented on lib/ddtrace/transport/http/service.rb:
```diff
@@ -0,0 +1,31 @@
+    module HTTP
+      # An HTTP service which tracer data can be sent to.
+      # Opens and manages connections.
+      class Service
```
Should we rename to not confuse with the other "service" we use for traces/spans?

delner replied:
What other name would you suggest? Not sure if I have a better suggestion. Maybe `Server`?

brettlangdon replied:
hmmm... why do you ask such difficult questions...

Agent? Client?

delner replied:
There's already a `Client` and `Agent` might be a bit too specific. Basically `Service` is supposed to represent "an addressable process which hosts an API." I thought `Service` was pretty fitting since it's congruent with the idea of "service oriented architecture", but `Server` could work too.

brettlangdon replied:
In Python our "Client" and "API" are the same class, so we just call it "API".

"Service" could be ok. I really don't have a better suggestion.

delner replied:
This `Service` class has been superseded by the `Adapter` class; the adapter manages connectivity to processes/services, and defines the specific delivery mechanism for it (e.g. `Net::HTTP`, Unix sockets, etc.)

brettlangdon commented on lib/ddtrace/transport/http/adapters/test.rb:
```diff
@@ -0,0 +1,77 @@
+            :status
+
+          def initialize(use_buffer = false)
+            @use_buffer = buffer
```
```suggestion
            @use_buffer = use_buffer
```

?

delner commented:
Replaced usage of `Datadog::HTTPTransport` with `Datadog::Transport::HTTP`, and added deprecation warnings to `Datadog::HTTPTransport`. Once I can get CI passing, this is ready for a final review.

delner commented on spec/ddtrace/transport/http/benchmark_spec.rb:
```diff
@@ -0,0 +1,34 @@
+      Benchmark.bm do |x|
+        x.report('Datadog::HTTPTransport') do
+          iterations.times do
+            traces = get_test_traces(2)
```
Seems like a good idea.

pawelchcki commented on lib/ddtrace/contrib/http/circuit_breaker.rb:
```diff
@@ -5,24 +5,52 @@ module HTTP
-          # we don't want to trace our own call to the API (they use net/http)
-          # when we know the host & port (from the URI) we use it, else (most-likely
+        def should_skip_tracing?(req, address, port, tracer)
+          return true if datadog_http_request?(req, address, port, tracer)
```
I was thinking about this concept - and I wonder if safer, and lower overhead approach would be to make a copy of `Net::HTTP` before it has been instrumented and use that class as a transport mechanism for sending traces to the agent.

irb snippet showing the concept of creating a class copy and using its original copy without the changes:
```
2.4.1 :001 > class A; def x; puts 'x'; end; end
 => :x
2.4.1 :002 > B = A.dup
 => B
2.4.1 :003 > A.class_eval do def x; puts 'a'; end; end
 => :x
2.4.1 :004 > A.new.x
a
 => nil
2.4.1 :005 > B.new.x
x
```

Would it make sense to use this pattern here ? Or do you find some cases where that would be more complex than existing solution ?

delner replied:
It sounds complex.

Ultimately the issue with this entire circuit breaker is that the `Net::HTTP` integration is assuming what the transport is, and how it submits traces; there's no way it can or should know this. (This is a problem with the design in general, not your suggestion.)

The best solution and replacement for this would be a "don't trace" block that the transport can activate itself to disable tracing for its operations. This would do a few things: 1) eliminate this circuit breaker from `Net::HTTP` entirely, 2) allow circuit breaking to be applied generally against any instrumented code the transport might run through, not just `Net:HTTP`. I would like to add this functionality in a future PR.

pawelchcki replied:
>The best solution and replacement for this would be a "don't trace" block that the transport can activate itself to disable tracing for its operations.

Its definitely possible to implement and would also lead to a cleaner code that what is currently here, so seems reasonable 👍
